Fix SE-0268 implementation for removal of parse-time name lookup

slavapestov opened the pull request:
We have to type check the `didSet` body before looking for references to `oldValue`, otherwise this check stops working once the parser no longer performs name lookup. This requires a bit of refactoring to break circular dependencies, since type checking the `didSet` body can in turn query attributes of the underlying storage; we don't want those queries to depend on type checking the `didSet` body.

There is a slight behavior change here. If a property wrapper has a mutating getter and a non-mutating setter, and it is used by a property with a `didSet`, then clearly the wrapped property must have a mutating getter as well. However, there is the question of whether the wrapped property's _setter_ should be mutating. Now, the setter calls `didSet`, so if `didSet` references `oldValue`, the setter also has to be mutating as well, since we must call the getter to load `oldValue`.

SE-0268 introduced the new behavior where the load of `oldValue` is elided if `didSet` does not reference `oldValue`; it also made it so that the setter did not need to be mutating in this case. This meant Swift 5.3 had a source and binary compatibility break from Swift 5.2. It also introduced dependencies between the interface type of the setter and the body of the `didSet`, which made this behavior no longer implementable once parse-time name lookup is disabled, because now we have to type check the body of the `didSet` in order to determine if `oldValue` is referenced or not; simply looking at the un-type checked AST is not enough.

This PR changes it back to match the old Swift 5.2 behavior, so that the setter is mutable in this case regardless of whether `oldValue` needs to be loaded or not. Hopefully, properties with a mutating getter and non-mutating setter should be rare, property wrappers like this doubly so, and wrapped properties with a `didSet` even more so.

See the forum discussion here: https://forums.swift.org/t/clarify-behavior-of-se-0268-with-a-mutating-getter/40324

slavapestov committed:
AST: Change isInvalid() on an accessor to delegate to the storage

This helps avoid a request cycle from the isInvalid() call inside
performAbstractFuncDeclDiagnostics().

slavapestov committed:
Sema: Type-check didSet body before looking for usages of 'oldValue'

Looking at the unchecked body only worked because parse-time lookup
bound references to 'oldValue' to point at the right ParamDecl.

By type-checking the body first, the implementation of SE-0268 can
work when parse-time lookup is disabled.

slavapestov committed:
Sema: Don't compute isSimpleDidSet() from getInterfaceType()

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov committed:
Sema: Bail out early when computing overrides of didSet, willSet, unsafeAddress and unsafeMutableAddress

slavapestov committed:
Sema: Don't compute isSimpleDidSet() from getStorageImplInfo()

slavapestov committed:
Sema: Don't remove @objc attribute when evaluating IsObjCRequest

This was preventing us from emitting diagnostics if IsObjCRequest
was evaluated before AttributeChecker::visitObjCAttr().

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

theblixguy reviewed (approved):
Looks good to me, thank you!

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov committed:
Sema: Don't consider if 'didSet' references 'oldValue' when computing if setter is mutable

Previously this would depend on whether the 'didSet' referenced
'oldValue'. This introduces a request cycle once parse-time name
lookup is disabled, and was also an ABI break from Swift 5.2.

See https://forums.swift.org/t/clarify-behavior-of-se-0268-with-a-mutating-getter/40324.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please smoke test macOS
